cmake/modules/hunter_download_server_url.cmake

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) 2017 NeroBurner
+# All rights reserved.
+
+include(CMakeParseArguments) # cmake_parse_arguments
+
+include(hunter_status_print)
+include(hunter_internal_error)
+include(hunter_test_string_not_empty)
+
+# used gobal variables
+# - HUNTER_DOWNLOAD_SERVER
+#   - if given, a list of servers to download from
+
+# update download url to point to HUNTER_DOWNLOAD_SERVER
+# - PACKAGE:       name of the package to update the url for
+# - VERSION:       version of the package
+# - SHA1:          SHA1 of the package to download, only first 7 digits are used as archive-ID
+# - URL:           original URL for the package
+# - OUTPUT:        variable to output the updated URL
+function(hunter_download_server_url)
+  set(function_name "hunter_download_server_url")
+  set(function_synopsis "${function_name}(PACKAGE foo VERSION 0.1.2-p3 SHA1 somelongsha URL http://original_url.example.com OUTPUT variable)")
+
+  # parse arguments
+  set(one_value_args PACKAGE VERSION SHA1 URL OUTPUT)
+  cmake_parse_arguments(x "" "${one_value_args}" "" ${ARGV})
+
+  # No free arguments allowed
+  list(LENGTH x_UNPARSED_ARGUMENTS x_len)
+  if(NOT x_len EQUAL "0")
+    hunter_internal_error(
+      "'${function_name}' incorrect usage,"
+      " expected no free arguments '${x_UNPARSED_ARGUMENTS}'."
+      " Synopsis: ${function_synopsis}"
+    )
+  endif()
+
+  # check mandatory arguments
+  foreach(arg PACKAGE VERSION SHA1 URL OUTPUT)
+    string(COMPARE EQUAL "${x_${arg}}" "" is_empty)
+    if(is_empty)
+      hunter_internal_error(
+        "'${function_name}' incorrect usage,"
+        " option '${arg}' with one argument is mandatory."
+        " Synopsis: ${function_synopsis}"
+      )
+    endif()
+  endforeach()
+
+  # default to the given URL
+  set(hunter_package_new_url "${x_URL}")
+
+  set(hunter_main_url) # variable for main_url (x_URL if it matches one of the download server)
+  set(hunter_url_list) # list of URLs to try downloading from
+
+  # check if download server is given
+  string(COMPARE NOTEQUAL "${HUNTER_DOWNLOAD_SERVER}" "" use_download_server)
```

Comment:
Since we are inside function you can use `return` to avoid indentation in next part:
```cmake
if(NOT use_download_server)
  set(${x_OUTPUT} ${hunter_package_new_url} PARENT_SCOPE)
  return()
endif()
```